src/lib/util/nest.mli

```diff
@@ -0,0 +1,60 @@
+(*     This file is distributed under the terms of OCamlPro               *)
+(*     Non-Commercial Purpose License, version 1.                         *)
+(*                                                                        *)
+(*     As an exception, Alt-Ergo Club members at the Gold level can       *)
+(*     use this file under the terms of the Apache Software License       *)
+(*     version 2.0.                                                       *)
+(*                                                                        *)
+(*     ---------------------------------------------------------------    *)
+(*                                                                        *)
+(*     The Alt-Ergo theorem prover                                        *)
+(*                                                                        *)
+(*     Sylvain Conchon, Evelyne Contejean, Francois Bobot                 *)
+(*     Mohamed Iguernelala, Stephane Lescuyer, Alain Mebsout              *)
+(*                                                                        *)
+(*     CNRS - INRIA - Universite Paris Sud                                *)
+(*                                                                        *)
+(*     Until 2013, some parts of this code were released under            *)
+(*     the Apache Software License version 2.0.                           *)
+(*                                                                        *)
+(*     ---------------------------------------------------------------    *)
+(*                                                                        *)
+(*     More details can be found in the directory licenses/               *)
+(*                                                                        *)
+(**************************************************************************)
+
+module DE = Dolmen.Std.Expr
+
+(** The purpose of this module is to construct a total order on ADT
+    constructors to ensure the termination of model generation for
+    this theory.
+
+    By nest, we mean the set of all the constructors in a mutually
+    recursive definition of ADTs.
+
+    Before comparing constructors with [compare], the complete
+    nest to which they belong have to be added with [add_nest]. *)
+
+type t = DE.ty_def list
+(** Type of nest. *)
+
+val add_nest : t -> unit
+(** [add_nest defs] sorts and adds the nest [defs] in the current
+    internal order.
+
+    {b Note:} Assume that [defs] is a complete nest (in any order). *)
+
+val compare : Uid.t -> Uid.t -> int
+(** [compare id1 id2] compares the constructors [id1] and [id2] with
+    the order generated by [add_nest] if they are both Dolmen identifiers.
+
+    @raise Failure if the nests of the constructors [id1] and [id2] haven't
+                   been added with [add_nest] before. *)
+
+val reinit : unit -> unit
+(** [reinit ()] resets the internal order. *)
```

Comment:
@bclement-ocp  I wrote this function for Alt-Ergo-Fuzz. The `Failure` in `compare` is there because we expect that the module Nest sorts all the constructors of a nest. If we use this function without declaring again nest, it will raise the failure exception. I think we should remove this function completely.

Reply:
I am fine with keeping this function — but the doc should clarify something along the line of "This clears all registered constructors." because it is not clear what "resets the internal order" means.